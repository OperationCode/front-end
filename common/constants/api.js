import { get, post, patch } from 'common/utils/api-utils';

/* GET REQUESTS */
export const getCodeSchoolsPromise = () => get('code_schools');
export const getMentorPromise = id => get(`mentors/${id}`);
export const getMentorsPromise = () => get('mentors');
export const getRequestsPromise = () => get('requests');
export const getScholarshipPromise = id => get(`scholarships/${id}`);
export const getScholarshipsPromise = () => get('scholarships');
export const getServicesPromise = () => get('services');
export const getTeamMembersPromise = () => get('team_members');

/* POST REQUESTS */
export const createUser = ({ email, password, firstName, lastName, zipcode }) =>
  post('users', {
    user: {
      email,
      password,
      first_name: firstName,
      last_name: lastName,
      zip: zipcode,
    },
  }).then(({ data }) => data);

export const loginUser = ({ email, password }) =>
  post('sessions', {
    user: {
      email,
      password,
    },
  }).then(({ data }) => data);

export const postMentorRequestPromise = ({ language, additionalDetails, mentor, service }) =>
  post('requests', {
    request: {
      details: additionalDetails,
      language,
      requested_mentor_id: mentor,
      service_id: service,
    },
  });

/* PATCH REQUESTS */
export const patchUpdateMentorRequestPromise = ({ request, status, mentor }) =>
  patch(`requests/${request}`, {
    request: {
      status,
      mentor,
    },
  });
