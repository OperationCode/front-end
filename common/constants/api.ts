import { patch, post } from 'common/utils/api-utils';
import type { RegistrationFormValues } from 'components/Forms/RegistrationForm/RegistrationForm';
import type { MilitaryStatusFormShape } from 'components/Forms/UpdateProfileForm/steps/MilitaryStatus';
import type { ProfessionalDetailsFormShape } from 'components/Forms/UpdateProfileForm/steps/ProfessionalDetails';

export const registerUser = (values: RegistrationFormValues) =>
  post('/api/registration/new', values);

export const updateUser = (values: ProfessionalDetailsFormShape | MilitaryStatusFormShape) =>
  patch('/api/registration/update', values);
