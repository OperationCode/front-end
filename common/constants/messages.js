export const validationErrorMessages = {
  required: 'Required',
  email: 'Must be a valid email',
  emailsMatch: 'Emails must match',
  emailExists: 'A user is already registered with this e-mail address',
  password:
    'Minimum password requirements: 8 characters, 1 Lowercase Letter, 1 Uppercase Letter, 1 Number',
  passwordsMatch: 'Passwords must match',
};

export const networkErrorMessages = {
  serverDown: 'Something is wrong on our end. Please try again later.',
};
