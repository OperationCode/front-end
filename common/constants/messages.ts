export const validationErrorMessages = {
  required: 'Required',
  email: 'Must be a valid email',
  emailsMatch: 'Emails must match',
  emailExists: 'A user is already registered with this e-mail address',
  codeOfConduct: 'You must agree to our Code of Conduct',
  slackGuidelines: 'You must agree to our Slack workspace guidelines',
};

export const networkErrorMessages = {
  serverDown: 'Something is wrong on our end. Please try again later.',
};
