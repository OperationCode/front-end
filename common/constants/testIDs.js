export const ACCORDION_CONTENT = 'ACCORDION_CONTENT';
export const ACCORDION_TOGGLE_BUTTON = 'ACCORDION_TOGGLE_BUTTON';
export const ALERT = 'ALERT';
export const ALERT_CLOSE_BUTTON = 'CLOSE_ALERT_BUTTION';
export const BUTTON = 'BUTTON';
export const CHANGE_PASSWORD_FORM_ERROR = 'CHANGE_PASSWORD_FORM_ERROR';
export const CLOSE_BUTTON = 'CLOSE_BUTTON';
export const LOGIN_BUTTON = 'LOGIN_BUTTON';
export const DOWNVOTE_BUTTON = 'DOWNVOTE_BUTTON';
export const FLAT_CARD_IMAGE = 'FLAT_CARD_IMAGE';
export const INPUT = 'INPUT';
export const INPUT_ERROR = 'INPUT_ERROR';
export const INPUT_FEEDBACK_GROUPING = 'INPUT_FEEDBACK_GROUPING';
export const LABEL = 'LABEL';
export const MULTI_STEP_SUBMIT_BUTTON = 'SUBMIT_MULTI_STEP_FORM';
export const MULTI_STEP_STEP_BUTTON = 'SUBMIT_STEP_BUTTON';
export const MULTI_STEP_PREVIOUS_BUTTON = 'MULTI_STEP_PREVIOUS_BUTTON';
export const SCREEN_READER_ONLY = 'SCREEN_READER_ONLY';
export const UPVOTE_BUTTON = 'UPVOTE_BUTTON';
export const RESOURCE_CARD = 'RESOURCE_CARD';
export const RESOURCE_TITLE = 'RESOURCE_TITLE';
export const RESOURCE_SEARCH = 'RESOURCE_SEARCH';
export const SKELETON_CARD = 'SKELETON_CARD';
export const RESOURCE_SEARCH_BUTTON = 'RESOURCE_SEARCH_BUTTON';
export const RESOURCE_RESET_BUTTON = 'RESOURCE_RESET_BUTTON';
export const NEXT_PAGE_BUTTON = 'NEXT_PAGE_BUTTON';
export const PREV_PAGE_BUTTON = 'PREV_PAGE_BUTTON';
export const DATA_TEST_LANGUAGES = '[data-test-languages]';
export const DATA_TEST_CATEGORY = '[data-test-category]';
export const DATA_TEST_COST = '[data-test-ispaid]';
export const COST_SELECT = 'input#react-select-paid-input';
export const LANGUAGE_SELECT = 'input#react-select-languages-input';
export const CATEGORY_SELECT = 'input#react-select-category-input';
export const PROFILE_GREETING = '[PROFILE_GREETING]';
export const SCHOOL_CARD_LOCATION_LIST_MODAL_BUTTON = 'SCHOOL_CARD_LOCATION_LIST_MODAL_BUTTON';
export const SCHOOL_LOCATION_LIST_ITEM = 'SCHOOL_LOCATION_LIST_ITEM';
export const MODAL_CONTENT = 'MODAL_CONTENT';
