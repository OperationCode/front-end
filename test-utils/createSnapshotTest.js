/* eslint-env jest */
import renderer from 'react-test-renderer';

/**
 * Used to create fully-rendered snapshot test.
 * Useful for reusable component smoke tests.
 *
 * @export
 * @param {*} Component
 */
export default Component => {
  const tree = renderer.create(Component).toJSON();
  expect(tree).toMatchSnapshot();
};
