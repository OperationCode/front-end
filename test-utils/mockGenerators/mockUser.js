import faker from 'faker';
import mockPassword from './mockPassword';

/**
 * @description generate a mocked user object for registration tests
 *
 * @export
 * @param {string} [desiredEmail=''] User will be created with this email, if passed
 * @returns {Object.<string, string>}
 */
export default function mockUser({
  desiredEmail = '',
  passwordOptions = {
    hasMinimumLength: true,
    hasOneLowercaseChar: true,
    hasOneUppercaseChar: true,
    hasOneNumber: true,
  },
} = {}) {
  const firstName = faker.name.firstName();
  const lastName = faker.name.lastName();
  const email = desiredEmail || faker.internet.email(firstName, lastName, 'operationcode.org');
  const password = mockPassword({
    hasMinimumLength: passwordOptions.hasMinimumLength,
    hasOneLowercaseChar: passwordOptions.hasOneLowercaseChar,
    hasOneUppercaseChar: passwordOptions.hasOneUppercaseChar,
    hasOneNumber: passwordOptions.hasOneNumber,
  });
  const zipcode = `${faker.address.zipCode()}`; // force to be string

  const user = {
    email,
    'confirm-email': email,
    password,
    'confirm-password': password,
    firstName,
    lastName,
    zipcode,
  };

  return user;
}
