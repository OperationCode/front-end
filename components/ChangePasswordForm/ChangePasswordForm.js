import React, { useState } from 'react';
import { string, func, shape } from 'prop-types';
import { Field, Formik } from 'formik';
import * as Yup from 'yup';
import Button from 'components/Button/Button';
import Form from 'components/Form/Form';
import Input from 'components/Form/Input/Input';
import Alert from 'components/Alert/Alert';
import { CHANGE_PASSWORD_FORM_ERROR } from 'common/constants/testIDs';
import { validationErrorMessages } from 'common/constants/messages';
import { getServerErrorMessage } from 'common/utils/api-utils';
import { minimumPasswordLength } from 'common/constants/validations';
import { hasRequiredCharacters } from 'common/utils/validator-utils';
import styles from './ChangePasswordForm.module.css';

const passwordResetSubmitSchema = Yup.object().shape({
  newPassword1: Yup.string()
    .required(validationErrorMessages.required)
    .min(minimumPasswordLength, validationErrorMessages.password)
    .test('password-strength', validationErrorMessages.password, hasRequiredCharacters),
  newPassword2: Yup.string()
    .required(validationErrorMessages.required)
    .oneOf([Yup.ref('newPassword1')], validationErrorMessages.passwordsMatch),
});

ChangePasswordForm.propTypes = {
  onSubmit: func.isRequired,
  onSuccess: func.isRequired,
  initialValues: shape({
    email: string,
  }),
};

ChangePasswordForm.defaultProps = {
  initialValues: {
    newPassword1: '',
    newPassword2: '',
  },
};

/**
 * Form component used for changing a password either during a password reset
 * or standard change password request.
 */
function ChangePasswordForm({ onSubmit, onSuccess, initialValues }) {
  const [errorMessage, setErrorMessage] = useState('');

  const handleSubmit = async (values, actions) => {
    try {
      await onSubmit(values);
      actions.setSubmitting(false);
      actions.resetForm();

      await onSuccess();
    } catch (error) {
      actions.setSubmitting(false);
      setErrorMessage(getServerErrorMessage(error));
    }
  };

  return (
    <Formik
      initialValues={initialValues}
      onSubmit={handleSubmit}
      validationSchema={passwordResetSubmitSchema}
    >
      {({ isSubmitting }) => (
        <Form className={styles.PasswordResetSubmitForm}>
          <div className={styles.row}>
            <Field
              type="password"
              name="newPassword1"
              label="Password*"
              component={Input}
              disabled={isSubmitting}
              autoComplete="new-password"
            />

            <Field
              type="password"
              name="newPassword2"
              label="Confirm Password*"
              component={Input}
              disabled={isSubmitting}
              autoComplete="new-password"
            />

            {errorMessage && (
              <Alert data-testid={CHANGE_PASSWORD_FORM_ERROR} type="error">
                {errorMessage}
              </Alert>
            )}

            <Button
              className={styles.topMargin}
              type="submit"
              theme="secondary"
              disabled={isSubmitting}
            >
              Submit
            </Button>
          </div>
        </Form>
      )}
    </Formik>
  );
}

export default ChangePasswordForm;
