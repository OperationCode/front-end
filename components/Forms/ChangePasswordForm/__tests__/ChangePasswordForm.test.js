import React from 'react';
import { fireEvent, render, waitFor } from '@testing-library/react';
import { passwordResetSubmit } from 'common/constants/api';
import { validationErrorMessages } from 'common/constants/messages';
import { BUTTON, INPUT_ERROR, CHANGE_PASSWORD_FORM_ERROR } from 'common/constants/testIDs';
import createSnapshotTest from 'test-utils/createSnapshotTest';
import OperationCodeAPIMock from 'test-utils/mocks/apiMock';
import mockUser from 'test-utils/mockGenerators/mockUser';
import ChangePasswordForm from '../ChangePasswordForm';

beforeEach(() => {
  OperationCodeAPIMock.reset();
});

describe('ChangePasswordForm', () => {
  const requiredProps = {
    onSuccess: jest.fn(),
    onSubmit: jest.fn(),
  };

  it('should render with required props', () => {
    createSnapshotTest(<ChangePasswordForm {...requiredProps} />);
  });

  it('should display required error message when blurring past password input', async () => {
    const component = render(<ChangePasswordForm {...requiredProps} />);

    fireEvent.blur(component.queryByLabelText(/^Password/));

    const InputError = await component.findByTestId(INPUT_ERROR);

    expect(InputError.textContent).toStrictEqual(validationErrorMessages.required);
  });

  it('should show "invalid password" message when given invalid password', async () => {
    const stringWithNoCapital = 'sillypassword1';

    const component = render(
      <ChangePasswordForm
        {...requiredProps}
        initialValues={{ newPassword1: stringWithNoCapital }}
      />,
    );

    fireEvent.click(component.queryByTestId(BUTTON));

    const InputError = await component.findByTestId(INPUT_ERROR);

    expect(InputError.textContent).toStrictEqual(validationErrorMessages.password);
  });

  it('should submit with valid data in form', async () => {
    const user = mockUser();

    const successSpy = jest.fn();
    const passwordResetSubmitSpy = jest.fn();

    const component = render(
      <ChangePasswordForm
        onSuccess={successSpy}
        onSubmit={passwordResetSubmitSpy}
        initialValues={{
          newPassword1: user.password,
          newPassword2: user.password,
        }}
      />,
    );

    expect(passwordResetSubmitSpy).not.toHaveBeenCalled();
    expect(successSpy).not.toHaveBeenCalled();

    fireEvent.click(component.queryByTestId(BUTTON));

    await waitFor(() => {
      expect(passwordResetSubmitSpy).toHaveBeenCalledTimes(1);
      expect(successSpy).toHaveBeenCalledTimes(1);
    });
  });

  it('should NOT submit to server with invalid data in form', async () => {
    const successSpy = jest.fn();
    const passwordResetSubmitSpy = jest.fn();

    const component = render(
      <ChangePasswordForm
        onSuccess={successSpy}
        onSubmit={passwordResetSubmitSpy}
        initialValues={{
          newPassword1: '1',
          newPassword2: '1',
        }}
      />,
    );

    fireEvent.click(component.queryByTestId(BUTTON));

    await waitFor(() => {
      expect(passwordResetSubmitSpy).not.toHaveBeenCalled();
      expect(successSpy).not.toHaveBeenCalled();
      expect(OperationCodeAPIMock.history.post.length).not.toBeGreaterThan(0);
    });
  });

  it('should display error message when request fails', async () => {
    const error = 'test error';
    const user = mockUser();

    OperationCodeAPIMock.onPost('auth/password/reset/confirm/', {
      newPassword1: user.password,
      newPassword2: user.password,
    }).reply(400, { error });

    const successSpy = jest.fn();

    const component = render(
      <ChangePasswordForm
        onSuccess={successSpy}
        onSubmit={passwordResetSubmit}
        initialValues={{
          newPassword1: user.password,
          newPassword2: user.password,
        }}
      />,
    );

    fireEvent.click(component.queryByTestId(BUTTON));

    await waitFor(() => {
      expect(OperationCodeAPIMock.history.post.length).toBeGreaterThan(0);
    });

    const FormError = await component.findByTestId(CHANGE_PASSWORD_FORM_ERROR);

    expect(FormError.textContent).toStrictEqual(error);
  });
});
