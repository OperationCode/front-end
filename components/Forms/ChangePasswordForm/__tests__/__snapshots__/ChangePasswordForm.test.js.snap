// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChangePasswordForm should render with required props 1`] = `
<form
  className="w-full"
  noValidate={true}
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="flex flex-col items-center"
  >
    <div
      className="field"
      data-testid="INPUT"
    >
      <label
        className="text-secondary text-sm"
        data-testid="LABEL"
        htmlFor="newPassword1"
      >
        Password*
      </label>
      <div
        className="inputFeedbackGrouping"
        data-testid="INPUT_FEEDBACK_GROUPING"
      >
        <input
          autoComplete="new-password"
          className="Input"
          disabled={false}
          id="newPassword1"
          name="newPassword1"
          onBlur={[Function]}
          onChange={[Function]}
          type="password"
          value=""
        />
      </div>
    </div>
    <div
      className="field"
      data-testid="INPUT"
    >
      <label
        className="text-secondary text-sm"
        data-testid="LABEL"
        htmlFor="newPassword2"
      >
        Confirm Password*
      </label>
      <div
        className="inputFeedbackGrouping"
        data-testid="INPUT_FEEDBACK_GROUPING"
      >
        <input
          autoComplete="new-password"
          className="Input"
          disabled={false}
          id="newPassword2"
          name="newPassword2"
          onBlur={[Function]}
          onChange={[Function]}
          type="password"
          value=""
        />
      </div>
    </div>
    <button
      className="Button mt-4 secondary"
      data-testid="BUTTON"
      disabled={false}
      onClick={[Function]}
      tabIndex={0}
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;
